### PR TITLE
Laboratory work #1, Polina Tokunova - 22FPL1

### PR DESCRIPTION
**Pull request by polsha24**

### CODE REVIEW

File: lab_1_classify_by_unigrams/start.py [hunk 1]
```diff
@@ -13,8 +14,9 @@ def main() -> None:
         de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+    result = create_language_profile("en", en_text)
     assert result, "Detection result is None"
+    print(result)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:19]: Remove, please

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -4,33 +4,43 @@
...
-    """
-    Creates a language profile
-    :param language: a language
-    :param text: a text
-    :return: a dictionary with two keys – name, freq
-    """
-
-
-def calculate_mse(predicted: list, actual: list) -> float | None:
+def tokenize(text: str) -> list[str] or None:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:7]: You must not change interface of functions and dock strings - please, revert!

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -4,33 +4,43 @@
...
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:8]: Try to use early return - code will be more clear

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -4,33 +4,43 @@
...
+        tokens = []
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:9]: use list comprehension

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -4,33 +4,43 @@
...
+        for symbol in text:
+            if symbol.isalpha():
+                tokens.append(symbol)
+        return tokens
+    else:
+        return None
+
+
+def calculate_frequencies(tokens: list[str] or None) -> dict[str, float] or None:
+    if isinstance(tokens, list) and all(isinstance(el, str) for el in tokens):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:20]: early return

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -4,33 +4,43 @@
...
+                freqs[token] = 1
+        for k, v in freqs.items():
+            freqs[k] = v / len(tokens)
+        return freqs
+    else:
+        return None
+
+
+def create_language_profile(language: str, text: str) -> dict[str, str or dict[str, float]] or None:
+    if isinstance(language, str) and isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:35]: again the same

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -40,9 +50,9 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
 
 
 def compare_profiles(
-        unknown_profile: dict[str, str | dict[str, float]],
-        profile_to_compare: dict[str, str | dict[str, float]]
-) -> float | None:
+        unknown_profile: dict[str, str or dict[str, float]],
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:53]: No, discard this changes

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Interesting start, please, fix review and continue with the next functions

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -5,29 +5,39 @@
...
 def tokenize(text: str) -> list[str] | None:
-    """
-    Splits a text into tokens, converts the tokens into lowercase,
-    removes punctuation, digits and other symbols
-    :param text: a text
-    :return: a list of lower-cased tokens without punctuation
-    """
+    if not isinstance(text, str):
+        return None
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:10]: you do not need `else`

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -5,29 +5,39 @@
...
 
 def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
-    """
-    Calculates frequencies of given tokens
-    :param tokens: a list of tokens
-    :return: a dictionary with frequencies
-    """
+    if not isinstance(tokens, list) or not all(isinstance(el, str) for el in tokens):
+        return None
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:22]: again

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -5,29 +5,39 @@
...
 def create_language_profile(language: str, text: str) -> dict[str, str | dict[str, float]] | None:
-    """
-    Creates a language profile
-    :param language: a language
-    :param text: a text
-    :return: a dictionary with two keys – name, freq
-    """
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+    else:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:37]: .

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -100,7 +110,7 @@ def detect_language_advanced(unknown_profile: dict[str, str | dict[str, float]],
     """
 
 
-def print_report(detections: list[tuple[str, float]]) -> None:
+def print_report(detections: list[list[str | float]]) -> None:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:113]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Revert all docstrings

File: lab_1_classify_by_unigrams/start.py (hunk 1 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/start.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -11,6 +12,11 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if not isinstance(text, str):
+        return None
+    text = text.lower()
+    tokens = [symbol for symbol in text if symbol.isalpha()]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:18]: return as soon as posible. Do not create the variable

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -28,6 +45,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     :param text: a text
     :return: a dictionary with two keys – name, freq
     """
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+    freqs = calculate_frequencies(tokenize(text))
+    profile = {"name": language, "freq": freqs}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:51]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -37,6 +59,14 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     :param actual: a list of actual values
     :return: the score
     """
+    if not isinstance(predicted, list) or not isinstance(actual, list) \
+            or (len(predicted) != len(actual)):
+        return None
+    dif = []
+    for i, value in enumerate(actual):
+        dif.append((value - predicted[i]) ** 2)
+    mse = sum(dif) / len(dif)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:68]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -49,6 +79,26 @@ def compare_profiles(
...
+    for symbol in all_symbols:
+        if symbol in unknown_profile.get("freq").keys():
+            unknown_freq.append(unknown_profile.get("freq").get(symbol))
+        else:
+            unknown_freq.append(0)
+        if symbol in profile_to_compare.get("freq").keys():
+            compare_freq.append(profile_to_compare.get("freq").get(symbol))
+        else:
+            compare_freq.append(0)
+    mse = calculate_mse(unknown_freq, compare_freq)
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:100]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Very good. You are so close!

File: requirements.txt [hunk 16]
```diff
@@ -1 +1,17 @@
+ast-comments==1.0.1
```
Comment (artyomtugaryov) [inline requirements.txt:1]: oh no:( revert!

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 16 above)
Comment (polsha24) [reply requirements.txt]: done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 17]
```diff
@@ -19,6 +24,17 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list) or not all(isinstance(el, str) for el in tokens):
+        return None
+    freqs = {}
+    for token in tokens:
+        if token in freqs:
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:31]: let's set 0 if key is not in the dict and increase the value always?

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -80,6 +147,17 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
...
+    if not isinstance(profile, dict) or "freq" not in profile.keys() \
+            or "name" not in profile.keys() or "n_words" not in profile.keys():
+        return None
+    profile_new = {"name": profile["name"], "freq": {}}
+    for token, freq in profile.get("freq").items():
+        if len(token) == 1 and re.search(r"\w", token):
+            if token.lower() in profile_new["freq"]:
+                profile_new["freq"][token.lower()] += freq / profile["n_words"][0]
+            else:
+                profile_new["freq"].update({token.lower(): freq / profile["n_words"][0]})
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:159]: `profile["n_words"][0]`, `token.lower()` -save to variables

File: lab_1_classify_by_unigrams/start.py [hunk 19]
```diff
@@ -1,20 +1,28 @@
...
+    profiles = collect_profiles(paths)
+    unknown_profile = create_language_profile("unknown", unknown_text)
+    if not isinstance(profiles, list) or not isinstance(unknown_profile, dict):
+        return None
+    result = detect_language_advanced(unknown_profile, profiles)
+    if not result:
+        return None
+    print_report(result)
     assert result, "Detection result is None"
+    return None
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:25]: not needed

File: lab_1_classify_by_unigrams/start.py [hunk 20]
```diff
@@ -1,20 +1,28 @@
...
-        de_text = file_to_read_de.read()
     with open("assets/texts/unknown.txt", "r", encoding="utf-8") as file_to_read_unk:
         unknown_text = file_to_read_unk.read()
-    result = None
+    langs = ["de", "en", "es", "fr", "it", "ru", "tr"]
+    paths = [f"assets/profiles/{langs[i]}.json" for i in range(7)]
+    profiles = collect_profiles(paths)
+    unknown_profile = create_language_profile("unknown", unknown_text)
+    if not isinstance(profiles, list) or not isinstance(unknown_profile, dict):
+        return None
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/start.py:19]: just return

File: requirements.txt [hunk 21]
```diff
@@ -1 +0,0 @@
-
```
Comment (artyomtugaryov) [inline requirements.txt:1]: revert

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 21 above)
Comment (polsha24) [reply requirements.txt]: done

File: lab_1_classify_by_unigrams/start.py (hunk 20 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/start.py]: done

File: lab_1_classify_by_unigrams/start.py (hunk 19 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/start.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 17 above)
Comment (polsha24) [reply lab_1_classify_by_unigrams/main.py]: done

**Pull request closed without merging**